EIPS/eip-5792.md

```diff
@@ -35,33 +35,41 @@ We also define one capability expression for use with the fourth method, which f
 
 ### `wallet_sendCalls`
 
-Requests that a wallet submits a batch of calls. `from` and `chainId` are top-level properties rather than per-call properties because all calls should be sent from the same sender and on the same chain, identified by [EIP-155](./eip-155.md) integers expressed in hexadecimal notation. The items in the `calls` field are only those that are shared by all transaction types. Any other fields that a wallet may need to submit a transaction should be handled by the wallet.
+Requests that a wallet submits a batch of calls. `from` and `chainId` are top-level properties rather than per-call properties because all calls should be sent from the same sender and on the same chain, identified by [EIP-155](./eip-155.md) integers expressed in hexadecimal notation.
+The items in the `calls` field are simple `{to, data, value}` tuples.
 
 The capabilities field is how an app can communicate with a wallet about capabilities a wallet supports. For example, this is where an app can specify a paymaster service URL from which an [ERC-4337](./eip-4337.md) wallet can request a `paymasterAndData` input for a user operation.
 
+Each capability defined in the "capabilities" member can define global or call specific fields.
+These fields are set inside this capability's entry in the `capabilitiesData` object.
+Each entity in the `calls` field may contain an optional `capabilitiesData` object.
+This object allows the applications to attach a capability-specific metadata to individual calls.
+
 The wallet:
 
 * MUST send the calls in the order specified in the request
-* MUST send the calls on the same chain identified by the call's `chainId` property
+* MUST send the calls on the same chain identified by the request's `chainId` property
 * MUST NOT await for any calls to be finalized to complete the batch
 * MUST NOT send any calls from the request if the user rejects the request
 * MAY revert all calls if any call fails
 * MAY send all calls as part of one or more transactions, depending on wallet capability
 * SHOULD stop executing the calls if any call fails
 * MAY reject the request if the from address does not match the enabled account
 * MAY reject the request if one or more calls in the batch is expected to fail, when simulated sequentially
+* MUST reject the request if it contains a `capability` that is not supported by the wallet
 
 #### `wallet_sendCalls` RPC Specification
 
 ```typescript
 type SendCallsParams = {
   version: string;
   from: `0x${string}`;
+  chainId: `0x${string}` | undefined; // Hex chain id
   calls: {
     to?: `0x${string}` | undefined;
     data?: `0x${string}` | undefined;
     value?: `0x${string}` | undefined; // Hex value
-    chainId?: `0x${string}` | undefined; // Hex chain id
+    capabilitiesData?: Record<string, any> | undefined;
```

Comment:
hmm, I don't love how this kind of muddles the simplicity of the calls struct. Is the complaint that capabilities makes it hard to specify info on a per call basis. Could we get a better example of when this would be needed 🤔 

> One example could be an "optional flag" capability that would allow a Smart Contract Wallet to proceed with the batch in case of a revert

This should be possible with a general capability? Or you're saying one call in particular can revert?

Reply:
You don't have to implement it, if our wallet doesn't have any per-call capability. 
Also, an app doesn't have to know about it, unless it needs that capability.
The idea that if someone ever want to add per-call capability, it doesn't mess directly with the "call" structure, or with such params of another capability.

Reply:
> Is the complaint that capabilities makes it hard to specify info on a per call basis?

Yes, that pretty much sums it up. Of course, the `capability` could provide an array of details, which would have to match the `calls` array's length exactly and contain `null` values for calls that do not require extra parameters, which is functionally equivalent but seems like a much more quirky API to me.

> Or you're saying one call in particular can revert?

In my example, yes, I was talking about an "atomic" batch with a single "optional" call that should not revert the batch;
It is hard to reason about potential future additional examples, so these are a little hypothetical, but in my opinion, worth considering:
If a "per-call gas limit capability" is defined, it can be used to provide a per-call gas limit;
In case a "multi-chain capability" is defined, the 'capabilitiesData' field can be used to specify a per-call 'chainId' and L2-specific parameters, something like `gas_per_pubdata_limit`;
In case a "conditional call capability" is defined, it can provide some description of a condition to include or exclude a call from the batch;
etc.

Does this make sense? Or should dapps never need to go to the trouble of defining per-call parameters for a `sendCalls` request?